File: SQL Server 2012/Projects/C#/Chapter 23/ApClrObjects/ApClrObjects/Functions/GetBalanceDueSum.cs
//------------------------------------------------------------------------------
// <copyright file="CSSqlFunction.cs" company="Microsoft">
//     Copyright (c) Microsoft Corporation.  All rights reserved.
// </copyright>
//------------------------------------------------------------------------------
using System;
using System.Data;
using System.Data.SqlClient;
using System.Data.SqlTypes;
using Microsoft.SqlServer.Server;

public partial class UserDefinedFunctions
{
    [Microsoft.SqlServer.Server.SqlFunction(DataAccess=DataAccessKind.Read)]
    public static SqlMoney GetBalanceDueSum()
    {
        // Put your code here
        // open the Connection object for the context
        SqlConnection connection = new SqlConnection("Context connection=true");
        connection.Open();

        // create the Command object
        SqlCommand command = new SqlCommand();
        command.Connection = connection;
        command.CommandText =
            "SELECT SUM(InvoiceTotal - PaymentTotal - CreditTotal) " +
            "FROM Invoices " +
            "WHERE InvoiceTotal - PaymentTotal - CreditTotal > 0";

        // execute the query
        decimal balanceDueSum = Convert.ToDecimal(command.ExecuteScalar());

        // close the Connection object
        connection.Close();

        // return the value
        return balanceDueSum;
    }
}

File: SQL Server 2012/Projects/Visual Basic/Chapter 23/ApClrObjects/Functions/GetBalanceDueSum.vb
Imports System
Imports System.Data
Imports System.Data.SqlClient
Imports System.Data.SqlTypes
Imports Microsoft.SqlServer.Server

Partial Public Class UserDefinedFunctions
    <Microsoft.SqlServer.Server.SqlFunction(DataAccess:=DataAccessKind.Read)> _
    Public Shared Function GetBalanceDueSum() As SqlMoney
        ' Open the Connection object for the context
        Dim connection As New SqlConnection("Context connection=true")
        connection.Open()

        ' Create the Command object
        Dim command As New SqlCommand()
        command.Connection = connection
        command.CommandText =
            "SELECT SUM(InvoiceTotal - PaymentTotal - CreditTotal) " &
            "FROM Invoices " &
            "WHERE InvoiceTotal - PaymentTotal - CreditTotal > 0"

        ' Execute the query
        Dim balanceDueSum As Decimal = CDec(command.ExecuteScalar)

        ' Close the Connection object
        connection.Close()

        ' Return the value
        Return balanceDueSum
    End Function
End Class
